ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java

```diff
@@ -49,21 +73,42 @@ public long getSize() {
 
   @Override
   public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
-    Future<Long> future;
-    if (!channelOpen.get()) {
-      ClosedChannelException closedChannelException = new ClosedChannelException();
-      FutureResult<Long> futureResult = new FutureResult<Long>();
-      futureResult.done(0L, closedChannelException);
-      future = futureResult;
-      if (callback != null) {
-        callback.onCompletion(0L, closedChannelException);
+    ReadIntoCallbackWrapper callbackWrapper = new ReadIntoCallbackWrapper(callback);
+    contentLock.lock();
+    try {
+      if (!channelOpen.get()) {
+        callbackWrapper.invokeCallback(new ClosedChannelException());
+      } else if (!channelEmptied.compareAndSet(false, true)) {
+        throw new IllegalStateException("ReadableStreamChannel cannot be read more than once");
       }
-    } else if (!channelEmptied.compareAndSet(false, true)) {
-      throw new IllegalStateException("ReadableStreamChannel cannot be read more than once");
+      Iterator<ByteBuffer> bufferIterator = buffers.iterator();
+      while (bufferIterator.hasNext()) {
+        ByteBuffer buffer = bufferIterator.next();
+        writeContent(asyncWritableChannel, callbackWrapper, buffer);
+        bufferIterator.remove();
+      }
+    } finally {
+      contentLock.unlock();
+    }
+    return callbackWrapper.futureResult;
+  }
+
+  /**
+   * Writes the provided {@code content} to the given {@code writeChannel}.
+   * @param writeChannel the {@link AsyncWritableChannel} to write the {@code content} to.
+   * @param callbackWrapper the {@link ReadIntoCallbackWrapper} for the read operation.
+   * @param content the piece of {@link ByteBuffer} that needs to be written to the {@code writeChannel}.
+   */
+  private void writeContent(AsyncWritableChannel writeChannel, ReadIntoCallbackWrapper callbackWrapper,
+                            ByteBuffer content) {
+    ContentWriteCallback writeCallback;
+    if (content == null) {
```

Comment:
Should have thrown `NullPointerException`. It doesn't make logical sense to have null `ByteBuffer` instances inbetween.